Add Windows + update for v0.2.0

jorisvandenbossche opened the pull request:
See https://github.com/conda-forge/pyogrio-feedstock/issues/6

I don't have windows myself, so can't debug this locally, but just to see what it already gives here on CI.

<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

jorisvandenbossche committed:
Add Windows

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

jorisvandenbossche commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 3.21.4, conda-smithy 3.9.0, and conda-forge-pinning 2021.03.05.11.08.04

jorisvandenbossche commented:
Hmm, the log output is not giving much clues ..

<details>

```
Processing d:\bld\pyogrio_1614950468418\work
  Created temporary directory: C:\Users\VssAdministrator\AppData\Local\Temp\pip-req-build-md9g1kzj
  Added file:///D:/bld/pyogrio_1614950468418/work to build tracker 'C:\\Users\\VssAdministrator\\AppData\\Local\\Temp\\pip-req-tracker-pbpnjzqa'
    Created temporary directory: C:\Users\VssAdministrator\AppData\Local\Temp\pip-modern-metadata-4s9acswb
    abstract_dist.prepare_distribution_metadata(finder, build_isolation)
  File "%PREFIX%\lib\site-packages\pip\_internal\distributions\sdist.py", line 41, in prepare_distribution_metadata
    self.req.prepare_metadata()
  File "%PREFIX%\lib\site-packages\pip\_internal\req\req_install.py", line 549, in prepare_metadata
    self.metadata_directory = self._generate_metadata()
  File "%PREFIX%\lib\site-packages\pip\_internal\req\req_install.py", line 534, in _generate_metadata
    return generate_metadata(
  File "%PREFIX%\lib\site-packages\pip\_internal\operations\build\metadata.py", line 34, in generate_metadata
    distinfo_dir = backend.prepare_metadata_for_build_wheel(
  File "%PREFIX%\lib\site-packages\pip\_vendor\pep517\wrappers.py", line 193, in prepare_metadata_for_build_wheel
    return self._call_hook('prepare_metadata_for_build_wheel', {
  File "%PREFIX%\lib\site-packages\pip\_vendor\pep517\wrappers.py", line 274, in _call_hook
    self._subprocess_runner(
  File "%PREFIX%\lib\site-packages\pip\_internal\utils\subprocess.py", line 289, in runner
    call_subprocess(
  File "%PREFIX%\lib\site-packages\pip\_internal\utils\subprocess.py", line 258, in call_subprocess
    raise InstallationSubprocessError(proc.returncode, command_desc)
pip._internal.exceptions.InstallationSubprocessError: Command errored out with exit status 1: '%PREFIX%\python.exe' '%PREFIX%\lib\site-packages\pip\_vendor\pep517\_in_process.py' prepare_metadata_for_build_wheel 'C:\Users\VSSADM~1\AppData\Local\Temp\tmpsoxyx9ct' Check the logs for full command output.
Removed file:///D:/bld/pyogrio_1614950468418/work from build tracker 'C:\\Users\\VssAdministrator\\AppData\\Local\\Temp\\pip-req-tracker-pbpnjzqa'
Removed build tracker: 'C:\\Users\\VssAdministrator\\AppData\\Local\\Temp\\pip-req-tracker-pbpnjzqa'
Traceback (most recent call last):
  File "C:\Miniconda\Scripts\conda-build-script.py", line 10, in <module>
    sys.exit(main())
  File "C:\Miniconda\lib\site-packages\conda_build\cli\main_build.py", line 481, in main
    execute(sys.argv[1:])
  File "C:\Miniconda\lib\site-packages\conda_build\cli\main_build.py", line 472, in execute
    verify=args.verify, variants=args.variants, cache_dir=args.cache_dir)
  File "C:\Miniconda\lib\site-packages\conda_build\api.py", line 195, in build
    variants=variants
  File "C:\Miniconda\lib\site-packages\conda_build\build.py", line 3073, in build_tree
    notest=notest,
  File "C:\Miniconda\lib\site-packages\conda_build\build.py", line 2159, in build
    windows.build(m, build_file, stats=build_stats, provision_only=provision_only)
  File "C:\Miniconda\lib\site-packages\conda_build\windows.py", line 299, in build
    check_call_env(cmd, cwd=m.config.work_dir, stats=stats, rewrite_stdout_env=rewrite_env)
  File "C:\Miniconda\lib\site-packages\conda_build\utils.py", line 412, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "C:\Miniconda\lib\site-packages\conda_build\utils.py", line 392, in _func_defaulting_env_to_os_environ
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['cmd.exe', '/d', '/c', 'conda_build.bat']' returned non-zero exit status 1.
##[error]Cmd.exe exited with code '1'.
```

</details>

brendan-ward commented:
@jorisvandenbossche thanks for working on this!

I was thinking about trying to setup build & test w/ Github actions to see if that would help identify some of the build issues here (agree: the logs are pretty inscrutible).  Was also thinking of looking at the conda recipe for [Fiona](https://github.com/conda-forge/fiona-feedstock) to see if that yields any hints (unless you've already done so?)

jorisvandenbossche commented:
Fiona has a bld.bat script: https://github.com/conda-forge/fiona-feedstock/blob/master/recipe/bld.bat, but, it also has such a similar script for linux. While here that doesn't seem needed?

jorisvandenbossche commented:
Of course, fiona's `setup.py` is also different, so many potential differences.

brendan-ward commented:
@jorisvandenbossche update: I've been working on a CI setup including Windows in this [branch](https://github.com/brendan-ward/pyogrio/tree/add_gh_ci).  As part of that, I've brought the setup.py and other config closer in line with Fiona (in particular, loading DLLs on Windows with Python >= 3.8 has extra challenges).  There were a lot of things that didn't work properly for Windows based on our original setup which likely contributed to failures here.  

I'm still debugging why the DLLs load correctly when the module is loaded directly but fail badly when loaded from the tests.  Once we have Windows CI tests passing, we can try and make corresponding updates here.

jorisvandenbossche commented:
@brendan-ward do you want to cut a new release so we can update this?

brendan-ward commented:
@jorisvandenbossche  Just added a new release: https://github.com/brendan-ward/pyogrio/releases/tag/v0.2.0

Thanks for the help here!

jorisvandenbossche committed:
update version

jorisvandenbossche committed:
add build scripts

jorisvandenbossche commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 3.21.4, conda-smithy 3.9.0, and conda-forge-pinning 2021.04.01.12.45.45

jorisvandenbossche commented:
@brendan-ward thanks! I added an sdist to the release, and updating the PR

jorisvandenbossche committed:
fix version

jorisvandenbossche committed:
try no pep517 + verbose

jorisvandenbossche commented:
The latest edit gave green builds! (not fully sure why it is needed here, though ..)

jorisvandenbossche committed:
clean-up unix script

jorisvandenbossche commented:
Going to merge this, so we have packages for the 0.2.0 release, and we can test the windows packages in practice

ocefpaf commented on recipe/bld.bat:
```diff
@@ -0,0 +1,9 @@
+:: There is no `gdal-config` on Windows so we need to hardcode gdal's version.
+set GDAL_VERSION=3.1.4
```
Maybe you can use something like this https://github.com/conda-forge/rasterio-feedstock/blob/master/recipe/bld.bat

jorisvandenbossche replied:
Ah, thanks. I copied it from fiona's feedstock ;)

BTW, I had expected that it would build against GDAL 3.1, since that's what is in the global pinnings (https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/2726e85da555e84e8ae7a2b4a33876495d285c68/recipe/conda_build_config.yaml#L461), but it actually built against GDAL 3.2 (latest available version). Did I misunderstand something about the pinnings?

ocefpaf replied:
[The migration overrides the pin](https://github.com/conda-forge/pyogrio-feedstock/blob/master/.ci_support/migrations/gdal32.yaml). When the migration is done the pin is updated.

PS: we should updated fiona too!

jorisvandenbossche replied:
I see, thanks for the explanation!

jorisvandenbossche replied:
-> https://github.com/conda-forge/pyogrio-feedstock/pull/9

brendan-ward commented:
Thank you so much @jorisvandenbossche for getting this working for Windows!  👏
